File: paper2_synchronization/python/producer_consumer.py
"""
producer_consumer.py -- Python threading bounded-buffer Producer-Consumer
Usage: python3 producer_consumer.py --buf 10 --producers 2 --consumers 2 --duration 2

Paper 2: Comparative Performance Evaluation of Multithreaded Synchronization
"""
import threading, time, argparse, psutil, os

class BoundedBuffer:
    def __init__(self, size):
        self.size    = size
        self.buffer  = [None] * size
        self.in_idx  = 0
        self.out_idx = 0
        self.mutex   = threading.Lock()
        self.empty   = threading.Semaphore(size)
        self.full    = threading.Semaphore(0)

class Producer(threading.Thread):
    def __init__(self, buf, stop_event, pid):
        super().__init__(daemon=True)
        self.buf = buf; self.stop = stop_event; self.pid = pid; self.count = 0
    def run(self):
        item = 0
        while not self.stop.is_set():
            item += 1
            if self.buf.empty.acquire(timeout=0.01):
                with self.buf.mutex:
                    self.buf.buffer[self.buf.in_idx] = item
                    self.buf.in_idx = (self.buf.in_idx + 1) % self.buf.size
                self.buf.full.release()
                self.count += 1

class Consumer(threading.Thread):
    def __init__(self, buf, stop_event, cid):
        super().__init__(daemon=True)
        self.buf = buf; self.stop = stop_event; self.cid = cid; self.count = 0
    def run(self):
        while not self.stop.is_set():
            if self.buf.full.acquire(timeout=0.01):
                with self.buf.mutex:
                    _ = self.buf.buffer[self.buf.out_idx]
                    self.buf.out_idx = (self.buf.out_idx + 1) % self.buf.size
                self.buf.empty.release()
                self.count += 1

def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("--buf",       type=int, default=10)
    ap.add_argument("--producers", type=int, default=2)
    ap.add_argument("--consumers", type=int, default=2)
    ap.add_argument("--duration",  type=int, default=2)
    args = ap.parse_args()

    buf = BoundedBuffer(args.buf)
    stop = threading.Event()
    proc = psutil.Process(os.getpid())

    producers = [Producer(buf, stop, i) for i in range(args.producers)]
    consumers = [Consumer(buf, stop, i) for i in range(args.consumers)]

    t0 = time.perf_counter()
    cpu0 = proc.cpu_times()
    mem0 = proc.memory_info().rss

    for t in producers + consumers: t.start()
    time.sleep(args.duration)
    stop.set()
    for t in producers + consumers: t.join(timeout=1.0)

    elapsed = time.perf_counter() - t0
    cpu1 = proc.cpu_times()
    mem1 = proc.memory_info().rss
    total = sum(p.count for p in producers)

    print(f"N={args.buf} M={args.producers} K={args.consumers} "
          f"elapsed={elapsed:.3f} produced={total} throughput={total/elapsed:.0f} "
          f"cpu_user={cpu1.user-cpu0.user:.3f} cpu_sys={cpu1.system-cpu0.system:.3f} "
          f"rss_mb={mem1/1e6:.1f}")

if __name__ == "__main__":
    main()

File: paper2_synchronization/c/producer_consumer.c
/*
 * producer_consumer.c -- POSIX pthreads bounded-buffer Producer-Consumer
 * Compile: gcc -O2 -Wall -o producer_consumer producer_consumer.c -lpthread -lrt
 * Usage:   ./producer_consumer --buf <N> --producers <M> --consumers <K> --duration <sec>
 *
 * Paper 2: Comparative Performance Evaluation of Multithreaded Synchronization
 */
#include <stdio.h>
#include <stdlib.h>
#include <pthread.h>
#include <semaphore.h>
#include <sys/resource.h>
#include <time.h>
#include <string.h>

typedef long buffer_item;

static buffer_item  *buffer;
static int           buf_size, buf_in = 0, buf_out = 0;
static pthread_mutex_t mutex;
static sem_t           sem_empty, sem_full;
static volatile int    running = 1;
static long           *produced_count, *consumed_count;
static int             num_producers, num_consumers;

static void *producer(void *arg) {
    int id = *(int *)arg;
    buffer_item item = 0;
    while (running) {
        item++;
        sem_wait(&sem_empty);
        pthread_mutex_lock(&mutex);
        buffer[buf_in] = item;
        buf_in = (buf_in + 1) % buf_size;
        produced_count[id]++;
        pthread_mutex_unlock(&mutex);
        sem_post(&sem_full);
    }
    return NULL;
}

static void *consumer(void *arg) {
    int id = *(int *)arg;
    buffer_item item;
    while (running) {
        sem_wait(&sem_full);
        pthread_mutex_lock(&mutex);
        item = buffer[buf_out];
        buf_out = (buf_out + 1) % buf_size;
        consumed_count[id]++;
        pthread_mutex_unlock(&mutex);
        sem_post(&sem_empty);
        (void)item;
    }
    return NULL;
}

int main(int argc, char **argv) {
    buf_size = 10; num_producers = 2; num_consumers = 2;
    int duration = 2;

    for (int i = 1; i < argc; i++) {
        if (!strcmp(argv[i], "--buf"))       buf_size      = atoi(argv[++i]);
        if (!strcmp(argv[i], "--producers")) num_producers = atoi(argv[++i]);
        if (!strcmp(argv[i], "--consumers")) num_consumers = atoi(argv[++i]);
        if (!strcmp(argv[i], "--duration"))  duration      = atoi(argv[++i]);
    }

    buffer         = calloc(buf_size, sizeof(buffer_item));
    produced_count = calloc(num_producers, sizeof(long));
    consumed_count = calloc(num_consumers, sizeof(long));

    pthread_mutex_init(&mutex, NULL);
    sem_init(&sem_empty, 0, buf_size);
    sem_init(&sem_full,  0, 0);

    pthread_t *prod_threads = malloc(num_producers * sizeof(pthread_t));
    pthread_t *cons_threads = malloc(num_consumers * sizeof(pthread_t));
    int *prod_ids = malloc(num_producers * sizeof(int));
    int *cons_ids = malloc(num_consumers * sizeof(int));

    struct timespec t_start, t_end;
    clock_gettime(CLOCK_MONOTONIC, &t_start);
    struct rusage ru_start;
    getrusage(RUSAGE_SELF, &ru_start);

    for (int i = 0; i < num_producers; i++) { prod_ids[i] = i; pthread_create(&prod_threads[i], NULL, producer, &prod_ids[i]); }
    for (int i = 0; i < num_consumers; i++) { cons_ids[i] = i; pthread_create(&cons_threads[i], NULL, consumer, &cons_ids[i]); }

    sleep(duration);
    running = 0;
    /* Unblock all blocked threads */
    for (int i = 0; i < num_producers; i++) sem_post(&sem_empty);
    for (int i = 0; i < num_consumers; i++) sem_post(&sem_full);

    for (int i = 0; i < num_producers; i++) pthread_join(prod_threads[i], NULL);
    for (int i = 0; i < num_consumers; i++) pthread_join(cons_threads[i], NULL);

    clock_gettime(CLOCK_MONOTONIC, &t_end);
    struct rusage ru_end;
    getrusage(RUSAGE_SELF, &ru_end);

    double elapsed = (t_end.tv_sec - t_start.tv_sec) + (t_end.tv_nsec - t_start.tv_nsec) / 1e9;
    long total_produced = 0;
    for (int i = 0; i < num_producers; i++) total_produced += produced_count[i];

    double cpu_user = (ru_end.ru_utime.tv_sec - ru_start.ru_utime.tv_sec)
                    + (ru_end.ru_utime.tv_usec - ru_start.ru_utime.tv_usec) / 1e6;
    double cpu_sys  = (ru_end.ru_stime.tv_sec - ru_start.ru_stime.tv_sec)
                    + (ru_end.ru_stime.tv_usec - ru_start.ru_stime.tv_usec) / 1e6;

    printf("N=%d M=%d K=%d elapsed=%.3f produced=%ld throughput=%.0f cpu_user=%.3f cpu_sys=%.3f nvcsw=%ld nivcsw=%ld rss=%ld\n",
           buf_size, num_producers, num_consumers, elapsed, total_produced,
           total_produced / elapsed, cpu_user, cpu_sys,
           ru_end.ru_nvcsw - ru_start.ru_nvcsw,
           ru_end.ru_nivcsw - ru_start.ru_nivcsw,
           ru_end.ru_maxrss);

    free(buffer); free(produced_count); free(consumed_count);
    free(prod_threads); free(cons_threads); free(prod_ids); free(cons_ids);
    return 0;
}
